RFCs/FS-1063-support-letbang-andbang-for-applicative-functors.md

```diff
@@ -0,0 +1,432 @@
+
+This may sound very constrained, but it is for good reason. The structure imposed by this rule forces the CE to be in a canonical form ([McBride & Paterson](http://www.staff.city.ac.uk/~ross/papers/Applicative.html)):
+
+> Any expression built from the Applicative combinators can be transformed to a canonical form in which a single pure function is "applied" to the effectful parts in depth-first order:  
+`pure f <*> arg1 <*> ... <*> argN`  
+This canonical form captures the essence of Applicative programming: computations have a fixed structure, given by the pure function, and a sequence of subcomputations, given by the effectful arguments.
+
+In our case, the expression to the right of `return` (i.e. `pure`) becomes the body of a lambda, whose parameters are introduced by the `let! ... and! ...` preceding it. This leads to a very straightforward desugaring of the syntax, which therefore ensures both using and understanding the feature is only as complex as is inherently required by the abstraction.
+
+Despite requiring the canonical form, there are still many ways to build more complex and useful expressions from this syntax. The rest of this section aims to give a tour around these various features.
+
+## Pattern Matching
+
+```fsharp
+let (|Quad|) (i : int) =
+    i * i * i * i
+
+type SingleCaseDu<'a> = SingleCaseDu of 'a
+
+ce {
+    let! Quad(x)          = foo
+    and! (y,_)            = bar
+    and! (SingleCaseDu z) = baz
+    return x + y + z
+ }
+```
+
+?
+
+```fsharp
+ce.Apply(
+    ce.Apply(
+        ce.Apply(
+            ce.Return(
+                (fun Quad(x) ->
+                    (fun (y,_) ->
+                        (fun (SingleCaseDu z) ->
+                            x + y + z
+                        )
+                    )
+                )),
+            foo),
+        bar), 
+    baz)
+```
+
+Pattern matching on the left-hand side of of a `let! ... and! ...` binding is valid, and desugars into the same pattern, but now as part of the lambda corresponding to that binding.
+
+## Using Monadic and Applicative Styles Simultaneously
+
+Recall that `let! ... and! ... return ...` syntax precludes an additional `let!` anywhere in the CE. In the case where your applicative happens also to be a monad, and you want to leverage the benefits of an applicative in some places (e.g. for performance reasons) but also use a `let!` (e.g. for convenience, or to do something a pure applicative doesn't support), you must do so inside a different CE context, e.g.:
+
+```fsharp
+ce {
+    let! quux =
+        ce {
+            let! x                = foo
+            and! (y,_)            = bar
+            and! (SingleCaseDu z) = baz
+            return x + y + z
+        }
+    if quux > 6
+    then
+        return quux
+    else
+        return 5
+}
+```
+
+?
```

Comment:
What is the significance of `?` here?

Reply:
These were arrows, but it looks like I accidentally corrupted them at some point. I'll fix that.